File: web/src/api/industryStrength.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export type IndustryState = 'EMERGING' | 'STRONG' | 'NEUTRAL' | 'COOLING' | 'WEAK';
export interface IndustrySnapshot {
  tradeDate: string; industryCode: string; industryName: string; state: IndustryState;
  close: number; ret1D: number; ret5D: number; ret10D: number; ret20D: number;
  rs5D: number; rs10D: number; rs20D: number; strengthScore: number; strengthRank: number;
  rankChange1D: number | null; rankChange3D: number | null; rankChange5D: number | null;
  previous5DReturn: number; momentumAcceleration5D: number; accelerationPercentile: number;
  turnoverRatio5D: number; constituentCount: number | null; dailyValidCount: number | null; ma5ValidCount: number | null; aboveMa5Count: number | null; ma20ValidCount: number | null; aboveMa20Count: number | null;
  upCount: number | null; downCount: number | null; flatCount: number | null; upRatio: number | null;
  aboveMa5Ratio: number | null; aboveMa20Ratio: number | null; equalWeightReturn: number | null;
  dataTimestamp: string; membersObservedAt: string | null; createdAt: string; updatedAt: string;
  quality: { breadthStatus?: string; dailyBreadthCoverage: number; ma5Coverage: number; ma20Coverage: number; catalogCount: number; rankedCount: number; coverage: number; excluded: Record<string, string> };
}
export interface IndustryRanking { tradeDate: string | null; expectedTradeDate: string; source: string; items: IndustrySnapshot[] }
export interface IndustryHistory { dates: string[]; items: IndustrySnapshot[] }
export interface IndustryDetail { current: IndustrySnapshot; history: IndustrySnapshot[] }
export interface Constituent {
  code: string; name: string; price: number | null; changePct: number | null; volume: number | null;
  amount: number | null; trendRank: number | null; aboveMa5: boolean | null; aboveMa20: boolean | null;
}
export interface Constituents { trendRankDate?: string | null; industryCode: string; updatedAt: string | null;
  constituentCount: number; dailyValidCount: number; ma5ValidCount: number; aboveMa5Count: number; ma20ValidCount: number; aboveMa20Count: number; items: Constituent[] }
export interface IndustryPreview {
  status: string; error: string | null;
  result: (IndustryRanking & { generatedAt: string; dataAsOf: string; dataLatestAt: string; constituents: Record<string, Constituents> }) | null;
}
const base = '/api/v1/industry-strength';
export const industryStrengthApi = {
  async preview(): Promise<IndustryPreview> {
    const result = toCamelCase<IndustryPreview>((await apiClient.get(`${base}/preview`)).data);
    if (result.result) {
      // Dynamic symbol keys must not be camel-cased (881101.TI -> 881101Ti).
      result.result.constituents = Object.fromEntries(
        Object.values(result.result.constituents).map(item => [item.industryCode, item]),
      );
    }
    return result;
  },
  async ranking(tradeDate?: string): Promise<IndustryRanking> {
    return toCamelCase((await apiClient.get(`${base}/ranking`, { params: { trade_date: tradeDate } })).data);
  },
  async dates(): Promise<string[]> { return (await apiClient.get(`${base}/dates`)).data; },
  async history(tradeDate?: string): Promise<IndustryHistory> {
    return toCamelCase((await apiClient.get(`${base}/history`, { params: { trade_date: tradeDate } })).data);
  },
  async detail(code: string, tradeDate?: string): Promise<IndustryDetail> {
    return toCamelCase((await apiClient.get(`${base}/${encodeURIComponent(code)}`, { params: { trade_date: tradeDate } })).data);
  },
  async constituents(code: string): Promise<Constituents> {
    return toCamelCase((await apiClient.get(`${base}/${encodeURIComponent(code)}/constituents`)).data);
  },
};

File: src/finance_analysis/interfaces/api/v1/schemas/industry_strength.py
"""Public, credential-free industry observation contracts (decimal returns)."""

from datetime import date, datetime
from typing import Any, Literal
from pydantic import BaseModel, Field

IndustryState = Literal["EMERGING", "STRONG", "NEUTRAL", "COOLING", "WEAK"]


class IndustrySnapshot(BaseModel):
    trade_date: date
    industry_code: str
    industry_name: str
    state: IndustryState
    data_timestamp: datetime
    members_observed_at: datetime
    created_at: datetime
    updated_at: datetime
    quality: dict[str, Any]
    close: float | None = None
    ret_1d: float | None = None
    ret_5d: float | None = None
    ret_10d: float | None = None
    ret_20d: float | None = None
    rs_5d: float | None = None
    rs_10d: float | None = None
    rs_20d: float | None = None
    rs_5d_percentile: float | None = None
    rs_10d_percentile: float | None = None
    rs_20d_percentile: float | None = None
    strength_score: float | None = None
    previous_5d_return: float | None = None
    momentum_acceleration_5d: float | None = None
    acceleration_percentile: float | None = None
    turnover_ratio_5d: float | None = None
    up_ratio: float | None = None
    above_ma5_ratio: float | None = None
    above_ma20_ratio: float | None = None
    equal_weight_return: float | None = None
    rs_5d_rank: int | None = None
    rs_10d_rank: int | None = None
    rs_20d_rank: int | None = None
    strength_rank: int | None = None
    rank_change_1d: int | None = None
    rank_change_3d: int | None = None
    rank_change_5d: int | None = None
    constituent_count: int | None = None
    valid_constituent_count: int | None = None
    up_count: int | None = None
    down_count: int | None = None
    flat_count: int | None = None


class RankingResponse(BaseModel):
    trade_date: date | None
    expected_trade_date: date
    source: str = "同花顺金融数据 API / 扶摇"
    items: list[IndustrySnapshot]


class HistoryResponse(BaseModel):
    dates: list[date]
    items: list[IndustrySnapshot]


class DetailResponse(BaseModel):
    current: IndustrySnapshot
    history: list[IndustrySnapshot]


class Constituent(BaseModel):
    code: str
    name: str
    price: float | None
    change_pct: float | None
    volume: float | None
    amount: float | None
    above_ma5: bool | None
    above_ma20: bool | None


class ConstituentsResponse(BaseModel):
    industry_code: str
    trade_date: date
    members_observed_at: datetime
    basis: str
    constituent_count: int
    valid_constituent_count: int
    up_count: int
    down_count: int
    flat_count: int
    up_ratio: float | None
    above_ma5_ratio: float | None
    above_ma20_ratio: float | None
    equal_weight_return: float | None
    items: list[Constituent] = Field(default_factory=list)
